omod/src/main/java/org/openmrs/module/spa/servlet/SpaServlet.java

```diff
@@ -55,10 +64,57 @@ protected long getLastModified(HttpServletRequest servletRequest) {
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
-        if (SpaModuleUtils.isRemoteAssetsEnabled()) {
-            handleRemoteAssets(request, response);
+        String requestURI = request.getRequestURI();
+        if (requestURI != null && requestURI.contains("frontend/config")) {
+            // get the config.json file
+            File configFile = getJsonConfigFile();
+            if (configFile.exists()) {
+                response.setContentType("application/json");
+                response.setHeader("Content-Disposition", "attachment; filename=\"" + configFile.getAbsolutePath() + "\"");
+
+                FileInputStream input = new FileInputStream(configFile);
+                byte[] buffer = new byte[1024];
+                int bytesRead;
+                while ((bytesRead = input.read(buffer)) != -1) {
+                    response.getOutputStream().write(buffer, 0, bytesRead);
+                }
+                input.close();
+            } else {
+                response.sendError(HttpServletResponse.SC_NOT_FOUND);
+            }
         } else {
-            handleLocalAssets(request, response);
+            if (SpaModuleUtils.isRemoteAssetsEnabled()) {
+                handleRemoteAssets(request, response);
+            } else {
+                handleLocalAssets(request, response);
+            }
+        }
+    }
+    @Override
+    protected void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+        String requestURI = request.getRequestURI();
+        if (requestURI.contains("frontend/config")) {
+
+            BufferedReader reader = request.getReader();
+            StringBuilder stringBuilder = new StringBuilder();
+            String line;
+
+            while ((line = reader.readLine()) != null) {
+                stringBuilder.append(line);
+            }
+
+            String requestBody = stringBuilder.toString();
+
+            File jsonConfigFile = getJsonConfigFile();
+            FileWriter writer = new FileWriter(jsonConfigFile);
+            writer.write(requestBody);
+            writer.close();
```

Comment:
Again, please just use the built in method for copying from one stream to another... We can just go from the reader to the writer... There's no reason to cache the whole thing in memory.

Reply:
Also, **never** close an InputStream or Writer without ensuring that it's done in a `finally` block or using the try-with-resources syntax (which does the same thing). If an exception occurs in the `writer.write()` call, the `close()` call will never be hit, which results in the application continuing to hold onto the underlying file descriptor for longer than it should.